control module / flight mode gpio library

cwalexand opened the pull request:
The control module and flight mode gpio library code is ready for review.

**Test mode guys:** To get us all started in the same place, a compatible blank gpio template has been provided for test mode. If you would prefer not to use it, I can remove it from the branch.

gpio testing was done on a raspberry pi with this:
https://github.com/hgb2/PSAS-Capstone/blob/flt_mode_libs/src/common/control.rs

The control module is based on this:
https://github.com/psas/reaction-control/blob/master/Controller_Scripts/Gain_v3.py

cwalexand committed:
added gpio libraries

cwalexand committed:
implemented the control module design

cwalexand committed:
added control module comments

cwalexand committed:
updated control module design

cwalexand committed:
updated flight mode gpio library design

BrianBreniser commented on src/flight_mode/src/gpio.rs:
```diff
@@ -1,8 +1,149 @@
+    io: Pin,
+}
+pub struct MyPins {
+    pins: LinkedList<Node>,
```
Jamie: Probably consider using Vec instead of LL

BrianBreniser commented on src/flight_mode/src/gpio.rs:
```diff
@@ -1,8 +1,149 @@
+/// 4) The requested GPIO is out of range and cannot be exported.
+/// 5) The requested GPIO is in use by the kernel and cannot
+///    be exported by use in userspace.
+pub fn add_pin(&mut self, pin_number: u64, direction: &str) {
```
Jamie: change direction to be a custom enum representing gpio directions

BrianBreniser replied:
Jamie: Consider not keeping the pin direction ourselves, the pin object should have that data already.

BrianBreniser commented on src/flight_mode/src/gpio.rs:
```diff
@@ -1,8 +1,149 @@
+        panic!("error exporting gpio pin {}: {}", pin_number, err);
+    }
+
+    if let Err(err) = node.io.set_direction(match direction.trim() {
```
Jamie: Try node.io = pin_dir.clone()

BrianBreniser commented on src/flight_mode/src/gpio.rs:
```diff
@@ -1,8 +1,149 @@
+use sysfs_gpio::{Pin, Direction};
+use std::collections::LinkedList;
+
+struct Node {
```
Jamie: Potentially remove the need to keep num/dir ourselves, use the Pin objects methods of accessing that data. Possibly remove Node altogether

BrianBreniser commented on src/common/control.rs:
```diff
@@ -1,17 +1,181 @@
+///////////////////////////////////////////////////////////////////////////////
+pub fn update(&mut self, mem: &mut SharedMemory) -> Result<u8, String> {
+    // stop_pin <- Get the value of the ESTOP pin
+    let stop_pin = match self.pins.get_value(ESTOP) {
```
Jamie: Check out Try! macro

BrianBreniser commented on src/common/control.rs:
```diff
@@ -1,17 +1,181 @@
+        // CALL write_pin(CW_pin, new state value, reference to shared memory)
+        self.state_update(rate_x);
+        let value = self.state;
+        if let Err(err) = self.write_pin(CW, value, mem) { return Err(err); }
```
Jamie: Try! macro again

BrianBreniser commented on src/common/control.rs:
```diff
@@ -1,17 +1,181 @@
+
+    let kp = 0.25;
+    let a = 2.0 * kp;
+    if rate_x < 0.0 {
```
Jamie: Find absolute value equivalent system for this.

BrianBreniser commented on src/common/control.rs:
```diff
@@ -1,17 +1,181 @@
+
+
+pub struct Control {
+    state: u8,
```
Jamie: Possibly change state to a bool, and convert by casting (in Rust: `as u8`)

BrianBreniser commented on src/common/control.rs:
```diff
@@ -1,17 +1,181 @@
+///////////////////////////////////////////////////////////////////////////////
+fn write_pin(&mut self, pin: u64, value: u8, mem: &mut SharedMemory) -> Result<(), String> {
+    // CALL pin.set_value(value)
+    if let Err(err) = self.pins.set_value(pin, value) {
```
Jamie: Try!

cwalexand committed:
removed compatible gpio template for test mode

cwalexand committed:
resolved code review issues for gpio.rs:

1) changed linked list to vector
2) created an enum for Direction
3) removed the Node structure that contained pin number and direction

cwalexand committed:
resolved code review issues for control.rs:

1) changed state variable to from u8 to bool
2) used try! macro where appropriate
3) used absolute value function call for rate_x

cwalexand committed:
updated flight mode gpio design to use generic container, not linked list
